### PR TITLE
LayoutResourceMgr

### PR DESCRIPTION
**Pull request by Pistonight**
All LayoutResourceMgr functions should match except for 3 that depend on `eui` which are undecompiled
Needs https://github.com/open-ead/sead/pull/143 for `sead::EnvUtil::getRegion`
Needs #129 for `nn::pl` - then the forward decl I added temporarily can be deleted

---
Also sorted `data_symbols.py` by address and added a script to do that

- - -
This change is [Reviewable](https://reviewable.io/reviews/zeldaret/botw/130)

### CODE REVIEW

**Comment by leoetlino** [comment]
sorry for the delay in reviewing this! could you rebase and fix the build failure?

**Comment by Pistonight** [comment]
> sorry for the delay in reviewing this! could you rebase and fix the build failure?

Need https://github.com/open-ead/nnheaders/pull/33 for one extra function, I will rebase again with the lib update

File: data/uking_functions.csv [hunk 1]
```diff
@@ -88328,31 +88328,31 @@ Address,Quality,Size,Name
...
+0x00000071010a28f0,U,000384,LayoutResourceMgr::loadGrammarArticleMaxLen
+0x00000071010a2a70,O,000260,_ZN4ksys2ui17LayoutResourceMgr12loadLangFontEPN4sead4HeapE
+0x00000071010a2b74,U,000388,LayoutResourceMgr::initializeFontMgr
+0x00000071010a2cf8,O,000680,_ZN4ksys2ui17LayoutResourceMgr18loadExtraLangFontsEPN4sead4HeapE
+0x00000071010a2fa0,O,000188,_ZN4ksys2ui17LayoutResourceMgr18checkLangFontReadyEv
+0x00000071010a305c,O,000136,_ZNK4ksys2ui17LayoutResourceMgr24checkExtraLangFontsReadyEv
+0x00000071010a30e4,O,000136,_ZN4ksys2ui17LayoutResourceMgr11loadVersionEv
+0x00000071010a316c,O,000372,_ZN4ksys2ui17LayoutResourceMgr17checkVersionReadyEv
+0x00000071010a32e0,U,001584,LayoutResourceMgr::doSomethingWithFontTextureCache
 0x00000071010a3910,W,000064,_ZN4sead12PerfJobQueue13measureEndRunEv
```
Comment (leoetlino) [inline data/uking_functions.csv:88345]: this seems mislabelled?

**Review by leoetlino — CHANGES_REQUESTED** [review]
> Reviewed 1 of 3 files at r6, 1 of 4 files at r7, 2 of 6 files at r11, 2 of 4 files at r12, 1 of 1 files at r16, 1 of 3 files at r17, 2 of 3 files at r18, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/zeldaret/botw/130)* status: 10 of 12 files reviewed, 10 unresolved discussions (waiting on @Pistonight)

___
*[`src/KingSystem/ksys.h` line 21 at r18](https://reviewable.io/reviews/zeldaret/botw/130#-O9CFLsB70sOOWwf-z4d:-O9CFLsB70sOOWwf-z4e:b-4s9nkd) ([raw file](https://github.com/zeldaret/botw/blob/1fdcd46494a8bd3f03be9491087f03306b005e31/src/KingSystem/ksys.h#L21)):*
> ```objc
> 
> // 0x0000007100f3a8d8
> void preInitializeApp(InitParams* params);
> ```

can this be a const ref instead of a pointer?
___
*[`src/KingSystem/ksys.cpp` line 35 at r18](https://reviewable.io/reviews/zeldaret/botw/130#-O9CF_pJ-mBTF9PlVOeG:-O9CF_pJ-mBTF9PlVOeH:b9splrf) ([raw file](https://github.com/zeldaret/botw/blob/1fdcd46494a8bd3f03be9491087f03306b005e31/src/KingSystem/ksys.cpp#L35)):*
> ```cpp
> 
> void preInitializeApp(InitParams* params) {
>     ksys::BasicProfiler::push("ksys::PreInitializeApp");
> ```

nit: could you use ksys::BasicProfiler::Scope instead of pushing/popping manually?
___
*[`src/KingSystem/System/UI/ArcResourceMgr.h` line 20 at r18](https://reviewable.io/reviews/zeldaret/botw/130#-O9CFrf66_rQ119D2eCg:-O9CFrf71Pbq1iMnzMJl:b-60wnhn) ([raw file](https://github.com/zeldaret/botw/blob/1fdcd46494a8bd3f03be9491087f03306b005e31/src/KingSystem/System/UI/ArcResourceMgr.h#L20)):*
> ```objc
> 
>     virtual void sub_10();
>     virtual void loadArchive(sead::ExpHeap* heap, const sead::SafeString& path);
> ```

are you sure this takes an ExpHeap and not a Heap? I'd be inclined to make this a Heap unless there's evidence that we need to use an ExpHeap (e.g. calling an ExpHeap member function).
___
*[`src/KingSystem/System/UI/LayoutResourceMgr.h` line 15 at r18](https://reviewable.io/reviews/zeldaret/botw/130#-O9CG5gZ6sjyNawavclt:-O9CG5gZ6sjyNawavclu:b-4nlgkp) ([raw file](https://github.com/zeldaret/botw/blob/1fdcd46494a8bd3f03be9491087f03306b005e31/src/KingSystem/System/UI/LayoutResourceMgr.h#L15)):*
> ```objc
> namespace ksys::ui {
> 
> class LayoutResourceMgr final : sead::hostio::Node {
> ```

public sead::hostio::Node
___
*[`src/KingSystem/System/UI/LayoutResourceMgr.h` line 96 at r18](https://reviewable.io/reviews/zeldaret/botw/130#-O9CGCgd-sHV2evkoKkP:-O9CGCgd-sHV2evkoKkQ:bfo1g6r) ([raw file](https://github.com/zeldaret/botw/blob/1fdcd46494a8bd3f03be9491087f03306b005e31/src/KingSystem/System/UI/LayoutResourceMgr.h#L96)):*
> ```objc
> };
> KSYS_CHECK_SIZE_NX150(LayoutResourceMgr, 0x168);
> constexpr const char* sExtraLangFontFiles[LayoutResourceMgr::cExtraLangCount *
> ```

can this be moved to the .cpp?
___
*[`src/KingSystem/System/UI/LayoutResourceMgr.cpp` line 156 at r18](https://reviewable.io/reviews/zeldaret/botw/130#-O9CHETC6RxtCVB-jFs5:-O9CHETC6RxtCVB-jFs6:bknou2b) ([raw file](https://github.com/zeldaret/botw/blob/1fdcd46494a8bd3f03be9491087f03306b005e31/src/KingSystem/System/UI/LayoutResourceMgr.cpp#L156)):*
> ```cpp
>     if (mVersionHandle->isReady()) {
>         auto* resource = sead::DynamicCast<sead::DirectResource>(mVersionHandle->getResource());
>         instance()->mVersionString.copy(reinterpret_cast<const char*>(resource->getRawData()),
> ```

wait what? they used the singleton instance pointer instead of just... setting mVersionString directly? lol
___
*[`src/KingSystem/System/UI/LayoutResourceMgr.cpp` line 214 at r18](https://reviewable.io/reviews/zeldaret/botw/130#-O9CHRlb6jmEVZq11ldM:-O9CHRlb6jmEVZq11ldN:b-rey02r) ([raw file](https://github.com/zeldaret/botw/blob/1fdcd46494a8bd3f03be9491087f03306b005e31/src/KingSystem/System/UI/LayoutResourceMgr.cpp#L214)):*
> ```cpp
> 
> bool LayoutResourceMgr::loadHorseLayoutResource() {
>     sead::ScopedLock<sead::CriticalSection> lock(&mCriticalSection);
> ```

nit: use sead::makeScopedLock so you don't have to write `sead::ScopedLock<sead::CriticalSection>`

(ideally this would be `sead::ScopedLock lock(...)` but CTAD is a C++17 feature that unfortunately isn't implemented in Clang 4.0...)
___
*[`src/KingSystem/System/UI/LayoutResourceMgr.cpp` line 228 at r18](https://reviewable.io/reviews/zeldaret/botw/130#-O9CHn0B3yRS_1Lf3MCk:-O9CHn0B3yRS_1Lf3MCl:bj3q5ko) ([raw file](https://github.com/zeldaret/botw/blob/1fdcd46494a8bd3f03be9491087f03306b005e31/src/KingSystem/System/UI/LayoutResourceMgr.cpp#L228)):*
> ```cpp
> 
> bool LayoutResourceMgr::unloadHorseLayout() {
>     sead::ScopedLock<sead::CriticalSection> lock(&mCriticalSection);
> ```

nit: sead::makeScopedLock
___
*[`src/KingSystem/System/UI/LayoutResourceMgr.cpp` line 240 at r18](https://reviewable.io/reviews/zeldaret/botw/130#-O9CHp_PBPuTaWcc7COn:-O9CHp_PBPuTaWcc7COo:b2fj3z0) ([raw file](https://github.com/zeldaret/botw/blob/1fdcd46494a8bd3f03be9491087f03306b005e31/src/KingSystem/System/UI/LayoutResourceMgr.cpp#L240)):*
> ```cpp
> 
> bool LayoutResourceMgr::loadArcResource(Archive& archive, const char* name) {
>     if (!archive.getResource()) {
> ```

nit: can we invert some of these if statements to return early?
___
*[`tools/format_data_symbols.py` line 13 at r18](https://reviewable.io/reviews/zeldaret/botw/130#-O9CHzW7AXHszXy92eC0:-O9CHzW7AXHszXy92eC1:bqvix3f) ([raw file](https://github.com/zeldaret/botw/blob/1fdcd46494a8bd3f03be9491087f03306b005e31/tools/format_data_symbols.py#L13)):*
> ```Python
>             if not line:
>                 continue
>             hex = "0x" + (line[2:18].upper())
> ```

the parsing is a bit rough here heh - consider using the (built-in) csv module?

File: data/uking_functions.csv (hunk 1 above)
Comment (Pistonight) [reply data/uking_functions.csv:88345]: Double-checked and it's not mislabelled

**Review by Pistonight — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/zeldaret/botw/130)* status: 5 of 12 files reviewed, 8 unresolved discussions (waiting on @leoetlino)

___
*[`src/KingSystem/System/UI/ArcResourceMgr.h` line 20 at r18](https://reviewable.io/reviews/zeldaret/botw/130#-O9CFrf66_rQ119D2eCg:-O9DfgdY9SpkA9tu_yRn:b1ajjff) ([raw file](https://github.com/zeldaret/botw/blob/1fdcd46494a8bd3f03be9491087f03306b005e31/src/KingSystem/System/UI/ArcResourceMgr.h#L20)):*
<details><summary><i>Previously, leoetlino (Léo Lam) wrote…</i></summary><blockquote>

are you sure this takes an ExpHeap and not a Heap? I'd be inclined to make this a Heap unless there's evidence that we need to use an ExpHeap (e.g. calling an ExpHeap member function).
</blockquote></details>

It's only used in new and assigning to sead::ResourceMgr::LoadArg. Will change to Heap
___
*[`src/KingSystem/System/UI/LayoutResourceMgr.h` line 15 at r18](https://reviewable.io/reviews/zeldaret/botw/130#-O9CG5gZ6sjyNawavclt:-O9DgB_wEXp0pl-2C5oT:b-896fix) ([raw file](https://github.com/zeldaret/botw/blob/1fdcd46494a8bd3f03be9491087f03306b005e31/src/KingSystem/System/UI/LayoutResourceMgr.h#L15)):*
<details><summary><i>Previously, leoetlino (Léo Lam) wrote…</i></summary><blockquote>

public sead::hostio::Node
</blockquote></details>

Done.
___
*[`src/KingSystem/System/UI/LayoutResourceMgr.h` line 96 at r18](https://reviewable.io/reviews/zeldaret/botw/130#-O9CGCgd-sHV2evkoKkP:-O9DgZq95jzC1bPQnJWy:b-896fix) ([raw file](https://github.com/zeldaret/botw/blob/1fdcd46494a8bd3f03be9491087f03306b005e31/src/KingSystem/System/UI/LayoutResourceMgr.h#L96)):*
<details><summary><i>Previously, leoetlino (Léo Lam) wrote…</i></summary><blockquote>

can this be moved to the .cpp?
</blockquote></details>

Done.
___
*[`src/KingSystem/System/UI/LayoutResourceMgr.cpp` line 156 at r18](https://reviewable.io/reviews/zeldaret/botw/130#-O9CHETC6RxtCVB-jFs5:-O9DgbNv0vB-t1Zkacmd:b-w43auv) ([raw file](https://github.com/zeldaret/botw/blob/1fdcd46494a8bd3f03be9491087f03306b005e31/src/KingSystem/System/UI/LayoutResourceMgr.cpp#L156)):*
<details><summary><i>Previously, leoetlino (Léo Lam) wrote…</i></summary><blockquote>

wait what? they used the singleton instance pointer instead of just... setting mVersionString directly? lol
</blockquote></details>

yes, I checked specifically using mVersionString does not match
___
*[`tools/format_data_symbols.py` line 13 at r18](https://reviewable.io/reviews/zeldaret/botw/130#-O9CHzW7AXHszXy92eC0:-O9DlT_gAZvYNq8l1hau:b-896fix) ([raw file](https://github.com/zeldaret/botw/blob/1fdcd46494a8bd3f03be9491087f03306b005e31/tools/format_data_symbols.py#L13)):*
<details><summary><i>Previously, leoetlino (Léo Lam) wrote…</i></summary><blockquote>

the parsing is a bit rough here heh - consider using the (built-in) csv module?
</blockquote></details>

Done.

**Review by Pistonight — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/zeldaret/botw/130)* status: 5 of 12 files reviewed, 7 unresolved discussions (waiting on @leoetlino)

___
*[`src/KingSystem/ksys.h` line 21 at r18](https://reviewable.io/reviews/zeldaret/botw/130#-O9CFLsB70sOOWwf-z4d:-O9DoI3n89KsBbkf7zC5:blpd809) ([raw file](https://github.com/zeldaret/botw/blob/1fdcd46494a8bd3f03be9491087f03306b005e31/src/KingSystem/ksys.h#L21)):*
<details><summary><i>Previously, leoetlino (Léo Lam) wrote…</i></summary><blockquote>

can this be a const ref instead of a pointer?
</blockquote></details>

Yeah, doesn't look like it's mutated in the function
___
*[`src/KingSystem/System/UI/LayoutResourceMgr.cpp` line 240 at r18](https://reviewable.io/reviews/zeldaret/botw/130#-O9CHp_PBPuTaWcc7COn:-O9DllF78DXuP29R7e0L:bbld5xu) ([raw file](https://github.com/zeldaret/botw/blob/1fdcd46494a8bd3f03be9491087f03306b005e31/src/KingSystem/System/UI/LayoutResourceMgr.cpp#L240)):*
<details><summary><i>Previously, leoetlino (Léo Lam) wrote…</i></summary><blockquote>

nit: can we invert some of these if statements to return early?
</blockquote></details>

Simplied 1 level of nesting

**Review by leoetlino — APPROVED** [review]
> Reviewed 1 of 3 files at r18, 5 of 5 files at r19, 3 of 3 files at r20, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/zeldaret/botw/130)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on @Pistonight)

**Pull request merged**